### PR TITLE
Feature/mraysu/add students form

### LINKED ISSUE #8: [ Add Students Form ] - Backend Route
We need a backend route in order to create a student and store the necessary data on MongoDb. The following fields that need to be stored are Student contact, Emergency Contact, Service Coordinator, Location, Birthdate, Medication, Dietary Restrictions, Intake-Date, Tour Date, Program Signed up for. See the [figma](https://www.figma.com/file/N16keN1mV1QDnorLB1GC6C/Plant-it-Again-(PIA)-Program-Manager?type=design&node-id=1034-6627&mode=design&t=arTX67jE53Uedjip-0) for more details on the information necessary.

References:

Screenshot 2024-01-10 at 12 40 43 AM

### PR DESCRIPTION
**Pull request by mraysu**
## Tracking Info

Resolves #8 

## Changes

Added student schema, validator, createStudent, post route

## Testing

Tested edge cases, missing fields, incorrect emails / phone numbers

## Confirmation of Change

See files in backend/src. Request bodies use following structure:
`{
"student":{ 
    "lastName":"Doe",
    "firstName":"John",
    "email":"bob@aol.com",
    "phoneNumber":"8005882300"
},
"emergency":{
    "lastName":"Doe",
    "firstName":"John",
    "email":"bob@aol.com",
    "phoneNumber":"8005882300"
},
"serviceCoordinator":{
    "lastName":"Doe",
    "firstName":"John",
    "email":"bob@aol.com",
    "phoneNumber":"8005882300"
},
"location":"HERE",
"medication":"ADVIL",
"birthday":"2023-10-01T00:00Z",
"intakeDate":"2023-10-01T00:00Z",
"tourDate":"2023-10-01T00:00Z",
"prog1":"Program 1",
"prog2":"Program 2",
"dietary":["seafood", "dairy"],
"otherString":"peanuts"
}`

### CODE REVIEW

File: backend/src/app.ts [hunk 1]
```diff
@@ -25,6 +27,8 @@ void mongoose
 
 server.app.use(json());
 
+server.app.use("/api/student", studentRoutes);
```
Comment (adhi0331) [inline backend/src/app.ts]: Let's have it be "/student"

File: backend/src/controllers/student.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    prog1,
+    prog2,
+    dietary,
+    otherString,
+  }: typedModel = req.body as typedModel; // ?????????
+
+  try {
+    let errornum = 0;
+    for (const error of errors.array()) {
+      errorHandler(new CustomError(errornum++, 400, error.msg as string), req, res);
```
Comment (adhi0331) [inline backend/src/controllers/student.ts]: Let's not use the CustomErrors in this way. We need to create the errors in the errors folders. See [this](https://github.com/TritonSE/ALUM-Mobile-Application/tree/main/backend/src/errors) as an example. We want to initialize the error handler like [this](https://github.com/TritonSE/ALUM-Mobile-Application/blob/main/backend/src/app.ts#L39). Store the errors in ../errors/vallidation.ts. Once this is setup properly we only need to throw that custom error.

File: backend/src/controllers/student.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+export const createStudent: RequestHandler = async (req, res, next) => {
+  type contact = {
+    lastName: string;
+    firstName: string;
+    email: string;
+    phoneNumber: string;
+  };
+
+  type typedModel = {
```
Comment (adhi0331) [inline backend/src/controllers/student.ts]: We should move these types outside of the route so that it can be accessed by future routes.

**Review by adhi0331 — CHANGES_REQUESTED** [review]
Great Work! The comments need to be addressed before I can approve.

**Review by adhi0331 — APPROVED** [review]
LGTM!!! 🚀

**Review by vs2961 — CHANGES_REQUESTED** [review]
I get the following error when I run the backend:
Screenshot 2024-01-27 at 12 19 17 PM

It could be an error on my end, but I'm unable to verify if this works or not.

**Review by aammya8 — CHANGES_REQUESTED** [review]
Looks good! Works for every test case I tested except the one below (where the user types in 0000000000 as their phone number). Not sure if we want to either (1) prevent the user from typing in 0000000000 as their phone number, or (2) store it as 0000000000, instead of 0 (which is how is is being stored now)

Screenshot 2024-01-27 at 11 47 43 PM

**Comment by mraysu** [comment]
In response to @vs2961's error, I am not reproduce the error. But, I looked into possible fixes and came across common issues with error middleware. [This thread](https://github.com/ladjs/supertest/issues/416#issuecomment-514508137) proposes a simple fix by adding the fourth "next" param to the function and cites the [expressjs documentation](https://expressjs.com/ru/api.html). We can see if this fix works, and possibly apply it to Aammya's branch to address @JacobLee23's very similar error.
image

**Comment by adhi0331** [comment]
@mraysu Victor and I were able to figure out what was causing the issue. Essentially, when you make a postman call to a route that has not been implemented (ex: localhost:4000/) then Express will call the error handler function and there error comes up. It isn't anything to do with your route. I'll resolve this bug so don't worry about it for now.

**Review by aammya8 — APPROVED** [review]

**Review by adhi0331 — APPROVED** [review]
LGTM!!!

**Review by vs2961 — APPROVED** [review]

**Pull request merged**